lib/api/route/messages.dart

```diff
@@ -1,11 +1,89 @@
 import 'package:json_annotation/json_annotation.dart';
 
 import '../core.dart';
+import '../exception.dart';
 import '../model/model.dart';
 import '../model/narrow.dart';
 
 part 'messages.g.dart';
 
+/// Convenience function to get a single message from any server.
+///
+/// This encapsulates a server-feature check.
+///
+/// Gives null if the server reports that the message doesn't exist.
+// TODO(server-5) Simplify this away; just use getMessage.
+Future<Message?> getMessageCompat(ApiConnection connection, {
+  required int messageId,
+  bool? applyMarkdown,
+}) async {
+  final useLegacyApi = connection.zulipFeatureLevel! < 120;
+  if (useLegacyApi) {
+    final response = await getMessages(connection,
+      narrow: [ApiNarrowMessageId(messageId)],
+      anchor: NumericAnchor(messageId),
+      numBefore: 0,
+      numAfter: 0,
+      applyMarkdown: applyMarkdown,
+
+      // Hard-code this param to `true`, as the new single-message API
+      // effectively does:
+      //   https://chat.zulip.org/#narrow/stream/378-api-design/topic/.60client_gravatar.60.20in.20.60messages.2F.7Bmessage_id.7D.60/near/1418337
+      clientGravatar: true,
+    );
+    final messages = response.messages;
+    if (messages.isEmpty) {
+      return null;
+    }
+    return messages.first;
+  } else {
+    try {
+      final response = await getMessage(connection,
+        messageId: messageId,
+        applyMarkdown: applyMarkdown,
+      );
+      // message property present at this feature level
+      return response.message!;
```

Comment:
This is probably fine in practice, but it's not the ideal behavior. I'd like to try to stick to the ideal behavior, to see if we have a pattern that makes it comfortable to do so.

The reason it's not ideal is that the `!` makes it effectively an internal error in our code if the property is missing. Whereas if it is missing, that signals instead that we have a malformed JSON response from the server — one that doesn't match our expectations from the API. Ordinarily, that would cause a MalformedServerResponseException, per this bit of `ApiConnection.send`:
```dart
    try {
      return fromJson(json);
    } catch (e) {
      throw MalformedServerResponseException(
        routeName: routeName, httpStatus: httpStatus, data: json);
    }
```

So this could instead check for null and throw such an exception in that case.

In this instance, though, I think there's actually a simpler way; see next comment.